### PR TITLE
add local feeds to protocol diagnostics

### LINKED ISSUE #8758: Protocol diagnostics: instrument local feeds

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: NuGet/Home#8758
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details:
Rename `ProtocolDiagnostics.Event` to `ProtocolDiagnostics.HttpEvent`.

Create `ProtocolDiagnostics.ResourceEvent`, triggered when `FindPackageById` or `DownloadResource` methods are called.

Create `ProtocolDiagnostics.NupkgCopiedEvent`, triggered when `IPackageDownloader.CopyNupkgFileToAsync` is called.

`PackageSourceTelemetry` aggregates all 3 of these events.

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

### CODE REVIEW

**Review by zkat — APPROVED** [review]

**Comment by zivkan** [comment]
I just remembered that I forgot to make the telemetry ignore global packages folder and fallback folder, so this PR isn't quite ready yet.

**Comment by zivkan** [comment]
This PR is ready for review now

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ public async Task<SearchResult<IPackageSearchMetadata>> SearchAsync(string searc
             }
 
             SearchResult<IPackageSearchMetadata> result;
-            using (var packageSourceTelemetry = new PackageSourceTelemetry(_sourceRepositories.Select(r => r.PackageSource), searchOperationId))
+            using (var packageSourceTelemetry = new PackageSourceTelemetry(_sourceRepositories, searchOperationId, "search"))
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs]: "search" should be a constant? Maybe per Feed or in some common area in PackageSourceTelemetry 
 or perhaps the PackageManagement namespace?

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs [hunk 2]
```diff
@@ -161,8 +161,8 @@ private async Task RestoreAsync(bool forceRestore, RestoreOperationSource restor
             _packageRestoreManager.PackageRestoredEvent += PackageRestoreManager_PackageRestored;
             _packageRestoreManager.PackageRestoreFailedEvent += PackageRestoreManager_PackageRestoreFailedEvent;
 
-            var sources = _sourceRepositoryProvider.GetRepositories().Select(s => s.PackageSource);
-            using (var packageSourceTelemetry = new PackageSourceTelemetry(sources, _nuGetProjectContext.OperationId))
+            var sources = _sourceRepositoryProvider.GetRepositories();
+            using (var packageSourceTelemetry = new PackageSourceTelemetry(sources, _nuGetProjectContext.OperationId, "restore"))
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs]: Another constant?

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 3]
```diff
@@ -5,130 +5,188 @@
...
+                }
+            }
+        }
 
-            var resourceData = pdEvent.Url.OriginalString.EndsWith(".nupkg", StringComparison.OrdinalIgnoreCase)
-                ? data.Nupkg
-                : data.Metadata;
+        private void ProtocolDiagnostics_HttpEvent(ProtocolDiagnosticHttpEvent pdEvent)
+        {
+            AddHttpData(pdEvent, _data, _knownSources);
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: Not following the purpose of having _data and _knownSources as parameters.
If they are properties, and the methods are non-static, why not just reference them directly?

These are also private/internal methods, so it's not like someone has the ability to pass in their own collections here.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 4]
```diff
@@ -5,130 +5,188 @@
...
+                return;
+            }
 
-                if (pdEvent.HeaderDuration.HasValue)
-                {
-                    if (resourceData.HeaderTiming == null)
-                    {
-                        resourceData.HeaderTiming = new ResourceTimingData();
-                    }
+            var data = allData.GetOrAdd(pdEvent.Source, _ => new Data());
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: Shouldn't it just be Get and NOT Add? If something gets added here, that means your sources (`_knownSources`) is now out of sync with allData (`_data`).

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 5]
```diff
@@ -5,130 +5,188 @@
...
             var parentId = _parentId.ToString();
             foreach (var kvp in _data)
             {
                 Data data = kvp.Value;
                 string source = kvp.Key;
-                if (!_sources.TryGetValue(kvp.Key, out PackageSource packageSource))
+                if (!_sources.TryGetValue(kvp.Key, out SourceRepository sourceRepository))
                 {
-                    packageSource = null;
+                    sourceRepository = new SourceRepository(new PackageSource(source), Repository.Provider.GetCoreV3());
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:201]: And update `_knownSources` ? Otherwise, seems subsequent Telemetry for the same source won't take advantage of the Dictionary.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 6]
```diff
@@ -148,113 +206,77 @@ internal static TelemetryEvent ToTelemetry(Data data, string source, PackageSour
...
+        {
+            telemetry[PropertyNames.Resources.Calls] = resources.Values.Sum(r => r.count);
+            telemetry.ComplexData[PropertyNames.Resources.Details] = ToResourceDetailsTelemetry(resources);
         }
 
-        private static void AddResourceProperties(TelemetryEvent telemetry, ResourceData data, PackageSourceDiagnosticsPropertyNames.ResourcePropertyNames propertyNames)
+        private static void AddHttpProperties(TelemetryEvent telemetry, HttpData data)
         {
-            lock (data.Lock)
+            telemetry[PropertyNames.Http.Requests] = data.Requests;
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:268]: Why remove the lock here? It's not possible for multiple threads to hit this at the same time?

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 7]
```diff
@@ -343,127 +363,82 @@ public Totals(int requests, long bytes, TimeSpan duration)
 
         internal class Data
         {
-            internal ResourceData Metadata { get; }
-            internal ResourceData Nupkg { get; }
+            internal object _lock;
+            internal Dictionary<string, (int count, TimeSpan duration)> Resources { get; }
+            internal HttpData Http { get; }
+            internal int NupkgCount { get; set; }
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:390]: Checking my understanding here... it looks like we used to track a Nupkg as a unit, but now we are merging the Count and Size of all Nupkgs into one metric. So we no longer can know the specific nupkg size that was being transferred, but only have the summed size?

File: src/NuGet.Core/NuGet.Protocol/LegacyFeed/DownloadResourceV2Feed.cs [hunk 8]
```diff
@@ -47,52 +57,65 @@ public DownloadResourceV2Feed(V2FeedParser feedParser)
...
+                        }
                     }
                 }
+                catch (OperationCanceledException)
+                {
+                    return new DownloadResourceResult(DownloadResourceResultStatus.Cancelled);
+                }
+                catch (Exception ex) when (!(ex is FatalProtocolException))
+                {
+                    // if the exception is not FatalProtocolException, catch it.
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Protocol/LegacyFeed/DownloadResourceV2Feed.cs]: // if the exception is not FatalProtocolException, `append some specific error message that's important for some reason`

Comment (donnie-msft) [reply src/NuGet.Core/NuGet.Protocol/LegacyFeed/DownloadResourceV2Feed.cs]: I see this already existed, but wonder if there's any point to this message?

File: src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        /// Initializes a new <see cref="LocalPackageArchiveDownloader" /> class.
+        /// </summary>
+        /// <param name="source">A package source.</param>
+        /// <param name="packageFilePath">A source package archive file path.</param>
+        /// <param name="packageIdentity">A package identity.</param>
+        /// <param name="logger">A logger.</param>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="packageFilePath" />
+        /// is either <c>null</c> or an empty string.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="packageIdentity" />
+        /// is either <c>null</c> or an empty string.</exception>
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs]: only if null, not empty string.

File: src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="logger" />
+        /// is either <c>null</c> or an empty string.</exception>
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs]: only if null, not empty string.

File: src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+        /// <summary>
+        /// Disposes of this instance.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!_isDisposed)
+            {
+                if (_packageReader.IsValueCreated)
+                {
+                    _packageReader.Value.Dispose();
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs:125]: When will this get disposed if IsValueCreated == false ?

File: src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+                }
+
+                if (_sourceStream.IsValueCreated)
+                {
+                    _sourceStream.Value.Dispose();
```
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs:130]: When will this get disposed if IsValueCreated == false ?

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 13]
```diff
@@ -104,38 +120,76 @@ public void AddAggregateData_TimingAggregation()
...
+            // Act
+            var resourceEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddResourceData(resourceEvent, data, stringTable)));
+            var httpEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddHttpData(httpEvent, data)));
+            var nupkgCopiedEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddNupkgCopiedData(nupkgCopiedEvent, data)));
+            await Task.WhenAll(resourceEvents, httpEvents, nupkgCopiedEvents);
 
             // Assert
             KeyValuePair<string, PackageSourceTelemetry.Data> pair = Assert.Single(data);
-            Assert.Equal(eventsToRaise, pair.Value.Nupkg.EventTiming.Requests);
+            Assert.Equal(eventsToRaise, pair.Value.Resources.Sum(r => r.Value.count));
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:173]: Not following how we are comparing the constant, 10,000, to the Sum of Resource's value's count.
Seems the more interesting test would be that the number of bytes (total nupkg size) got added up correctly.

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 14]
```diff
@@ -144,22 +198,21 @@ public void AddAggregateData_CountsBytes()
...
-            var nupkgData = pair.Value.Nupkg;
-            Assert.Equal(bytes.Sum(), nupkgData.TotalBytes);
-            Assert.Equal(bytes.Max(), nupkgData.MaxBytes);
+            var httpData = pair.Value.Http;
+            Assert.Equal(bytes.Sum(), httpData.TotalBytes);
         }
 
         [Fact]
-        public void AddAggregateData_Counts()
+        public void AddHttpData_MultipleEvents_CountsBools()
```
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:199]: This entire test feels weird to me. trueFalseList is only ever 1 true and 1 false? So what's the point of all the looping and division by 2. Not sure about the maintainability of this one, and it's certainly not intuitive to me, anyway.

**Review by donnie-msft — APPROVED** [review]
Nothing shocking here, just a few nits and seeking some clarifications.

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs [hunk 15]
```diff
@@ -44,9 +50,26 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
...
+        /// <summary>
+        /// Initializes a new <see cref="HttpFileSystemBasedFindPackageByIdResource" /> class.
+        /// </summary>
+        /// <param name="baseUris">Base URI's.</param>
+        /// <param name="httpSource">An HTTP source.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="baseUris" /> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">Thrown if <paramref name="baseUris" /> is empty.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="httpSource" /> is <c>null</c>.</exception>
+        public HttpFileSystemBasedFindPackageByIdResource(
+            string source,
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs]: Can the `source` argument be obtained from `httpSource` argument ? (i.e. `httpSource.PackageSource.Source`)

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 16]
```diff
@@ -21,81 +23,95 @@ public void Ctor_DuplicateSourceUrls_DoesNotThrow()
...
             // Arrange
-            var pde = CreateSampleProtocolDiagnosticsEvent(url: new Uri(url));
+            var re1 = CreateSampleResourceEvent(source: "source1");
+            var re2 = CreateSampleResourceEvent(source: "source2");
             var data = new ConcurrentDictionary<string, PackageSourceTelemetry.Data>();
+            var stringTable = new ConcurrentDictionary<string, ConcurrentDictionary<string, string>>();
 
             // Act
-            PackageSourceTelemetry.AddAggregateData(pde, data);
+            PackageSourceTelemetry.AddResourceData(re1, data, stringTable);
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs]: In this PR, `PackageSourceTelemetry.AddResourceData()` accepts 4 arguments. Here, you're using only 3.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 17]
```diff
@@ -5,130 +5,188 @@
...
 {
     public sealed class PackageSourceTelemetry : IDisposable
     {
         private readonly ConcurrentDictionary<string, Data> _data;
-        private readonly IDictionary<string, PackageSource> _sources;
+        private readonly IDictionary<string, SourceRepository> _sources;
         private readonly Guid _parentId;
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, string>> _resourceStringTable;
+        private readonly string _actionName;
+        private readonly HashSet<string> _knownSources;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: Do you still need _knownSources? Looks like it's covered by _sources.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs [hunk 18]
```diff
@@ -5,130 +5,188 @@
...
+using NuGet.Protocol.Core.Types;
 using NuGet.Protocol.Utility;
 
 namespace NuGet.VisualStudio.Telemetry
 {
     public sealed class PackageSourceTelemetry : IDisposable
     {
         private readonly ConcurrentDictionary<string, Data> _data;
-        private readonly IDictionary<string, PackageSource> _sources;
+        private readonly IDictionary<string, SourceRepository> _sources;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:21]: Why did you need to use a SourceRepository?
I see you are using PackageSource anyways.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs [hunk 19]
```diff
@@ -23,6 +23,8 @@
 
 namespace NuGet.Commands.Test
 {
+    using LocalPackageArchiveDownloader = NuGet.Protocol.LocalPackageArchiveDownloader;
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs:26]: The using statements sit outside of the namespace.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/NugetPackageUtilTests.cs [hunk 20]
```diff
@@ -16,6 +16,8 @@
 
 namespace Commands.Test
 {
+    using LocalPackageArchiveDownloader = NuGet.Protocol.LocalPackageArchiveDownloader;
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/NugetPackageUtilTests.cs:19]: The using statements sit outside of the namespace.

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageExtractorTests.cs [hunk 21]
```diff
@@ -27,6 +27,8 @@
 
 namespace NuGet.Packaging.Test
 {
+    using LocalPackageArchiveDownloader = NuGet.Protocol.LocalPackageArchiveDownloader;
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageExtractorTests.cs:30]: The using statements sit outside of the namespace.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs (hunk 19 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs:26]: Seeing your change below...is there an ordering issue?

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs [hunk 22]
```diff
@@ -715,7 +715,9 @@ private static string ResolveRelativePath(string originDirectoryPath, string ori
                 return path;
             }
 
-            return Path.Combine(originDirectoryPath, ResolvePath(Path.GetDirectoryName(originFilePath), path));
+            var rawPath = Path.Combine(originDirectoryPath, ResolvePath(Path.GetDirectoryName(originFilePath), path));
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:718]: Can you elaborate what scenario failed?

File: src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs [hunk 23]
```diff
@@ -15,6 +15,7 @@ namespace NuGet.Packaging
     /// <summary>
     /// A package downloader for local archive packages.
     /// </summary>
+    [Obsolete("Use NuGet.Protocol.LocalPackageArchiveDownloader")]
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/LocalPackageArchiveDownloader.cs:18]: 👏

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs (hunk 22 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:718]: Is originDirectoryPath not rooted?

File: src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs (hunk 11 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/LocalPackageArchiveDownloader.cs:125]: If it wasn't created, there's nothing to dispose.
This check is good because otherwise you'd initialize and dispose of the object.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceTests.cs [hunk 24]
```diff
@@ -27,6 +27,7 @@ public void Constructor_ThrowsForNullBaseUris()
         {
             var exception = Assert.Throws<ArgumentNullException>(
                 () => new HttpFileSystemBasedFindPackageByIdResource(
+                    source: "source",
```
Comment (dominoFire) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceTests.cs]: nit: If you consider keeping the source argument in the constructor, consider using a constant for the test `"source"` string

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs (hunk 15 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs]: agree with @dominoFire  It's already been done that way anyways, and there's a null check on the package source param :)

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalDownloadResourceTests.cs [hunk 25]
```diff
@@ -47,7 +47,7 @@ public async Task LocalDownloadResource_PackageIsReturnedAsync()
                 string packagesFolder = null; // This is unused by the implementation.
 
                 var localResource = new FindLocalPackagesResourceV2(root);
-                var resource = new LocalDownloadResource(localResource);
+                var resource = new LocalDownloadResource("source", localResource);
```
Comment (dominoFire) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/LocalResourceTests/LocalDownloadResourceTests.cs:50]: nit: Consider using a string constant and reference it through the tests.

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs [hunk 26]
```diff
@@ -161,24 +198,37 @@ public class HttpFileSystemBasedFindPackageByIdResource : FindPackageByIdResourc
...
 
-                return GetDependencyInfo(reader);
-            }
+                    return GetDependencyInfo(reader);
+                }
 
-            return null;
+                return null;
+            }
+            finally
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:204]: This might be tricky to reason about.
In this call is where pretty much all the downloads from V3 feeds happen, because we call the nupkg downloader.

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceProvider.cs [hunk 27]
```diff
@@ -33,6 +33,7 @@ public HttpFileSystemBasedFindPackageByIdResourceProvider()
                 var httpSourceResource = await sourceRepository.GetResourceAsync<HttpSourceResource>(token);
 
                 resource = new HttpFileSystemBasedFindPackageByIdResource(
+                    sourceRepository.PackageSource.Source,
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResourceProvider.cs]: revert this

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/RemoteV3FindPackageByIdResource.cs [hunk 28]
```diff
@@ -150,22 +168,35 @@ public RemoteV3FindPackageByIdResource(SourceRepository sourceRepository, HttpSo
...
+                    packageInfo.Identity,
+                    packageInfo.ContentUri,
+                    cacheContext,
+                    logger,
+                    cancellationToken);
 
-            return GetDependencyInfo(reader);
+                return GetDependencyInfo(reader);
+            }
+            finally
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/RemoteRepositories/RemoteV3FindPackageByIdResource.cs:191]: This will never be used, but anyways, this is where the actual download happens.

**Review by nkolev92 — COMMENTED** [review]
👏 Looks great overall, I'm just curious how we will be able to reason about the telemetry. 

Address mostly cleanup related feedback before approval.

**Comment by nkolev92** [comment]
Actually I see that we are still keeping the protocol diagnostics telemetry, so technically the measurements overlap, but they'll help paint a more complete picture.

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 29]
```diff
@@ -21,81 +23,95 @@ public void Ctor_DuplicateSourceUrls_DoesNotThrow()
...
+        public void AddHttpData_NullHeaderDuration_SetsResultToNull()
+        {
+            // Arrange
+            var pde1 = CreateSampleHttpEvent(headerDuration: TimeSpan.FromMilliseconds(150));
+            var pde2 = CreateSampleHttpEvent(headerDuration: null);
+            var pde3 = CreateSampleHttpEvent(headerDuration: TimeSpan.FromMilliseconds(100));
+            var data = new ConcurrentDictionary<string, PackageSourceTelemetry.Data>();
+
+            // Act
+            PackageSourceTelemetry.AddHttpData(pde1, data);
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:87]: `PackageSourceTelemetry.AddHttpData()` accepts 3 arguments. Here, you're using only 2.

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 30]
```diff
@@ -104,38 +120,76 @@ public void AddAggregateData_TimingAggregation()
...
+            var sizes = new[] { 1, 22, 333, 4444, 55555, 666666 };
             var data = new ConcurrentDictionary<string, PackageSourceTelemetry.Data>();
-            var eventsToRaise = 10000;
 
             // Act
-            Parallel.For(0, eventsToRaise, _ =>
+            for (int i = 0; i < sizes.Length; i++)
+            {
+                var nce = new ProtocolDiagnosticNupkgCopiedEvent("source", sizes[i]);
+                PackageSourceTelemetry.AddNupkgCopiedData(nce, data);
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:128]: `PackageSourceTelemetry.AddNupkgCopiedData()` accepts 3 arguments. Here, you're using only 2

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs [hunk 31]
```diff
@@ -104,38 +120,76 @@ public void AddAggregateData_TimingAggregation()
...
+
+                return Task.WhenAll(tasks);
+            }
+            var source = "http://source.test/v3/index.json";
+            var resourceEvent = CreateSampleResourceEvent(source: source);
+            var httpEvent = CreateSampleHttpEvent(source: source);
+            var nupkgCopiedEvent = new ProtocolDiagnosticNupkgCopiedEvent(source, fileSize: 123456);
+
+            // Act
+            var resourceEvents = Task.Run(() => SendEvents(() => PackageSourceTelemetry.AddResourceData(resourceEvent, data, stringTable)));
```
Comment (dominoFire) [inline test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:166]: Do you mean adding a sources argument for `PackageSourceTelemetry.AddResourceData()` ?

**Review by dominoFire — DISMISSED** [review]
Please address the comments in `PackageSourceTelemetry`

File: src/NuGet.Core/NuGet.Protocol/Utility/ProtocolDiagnosticNupkgCopiedEvent.cs [hunk 32]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Protocol.Utility
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Protocol/Utility/ProtocolDiagnosticNupkgCopiedEvent.cs]: It seems more appropriate that all of these telemetry types be moved into a Telemetry folder and the namespace (NuGet.Protocol.Telemetry.)

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 3 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: The class constructor makes global state changes (subscribes to a static event). I don't want tests to change or use global state, particularly so if tests run in parallel, there are no flaky failures. So, all the instance methods call static methods with the business logic, so that tests can call the static methods directly, without affecting any static values.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 4 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: `_data` wasn't pre-populated with data for all known sources, so `GetOrAdd` was needed. But then I realized I could just pre-populate it with all known sources, and remove the `_knownSources` hash set.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 6 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:268]: Previous code had two data objects per source, after the refactor there is one. It was intended that the `ToTelemetry` method gets the lock, but there was a bug which will be fixed in the next push.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 7 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:390]: `PackageSourceTelemetry` always aggregated/accumulated numbers, so unless there's only a single nupkg, we won't know the size of each individually.

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs (hunk 13 above)
Comment (zivkan) [reply test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:173]: discussed offline

File: test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs (hunk 14 above)
Comment (zivkan) [reply test/NuGet.Clients.Tests/NuGet.VisualStudio.Common.Test/Telemetry/PackageSourceTelemetryTests.cs:199]: discussed offline

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 17 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs]: `_knownSources` was intended for performance, the `ProtocolDiagnosticEvent` events will be called a large number of times, so the difference between a linear lookup for a `List` could be much slower than a quick `HashSet<>.Contains`.  However, from another comment realized that if I pre-populated the `_data`, then it can double as a fast known-source whitelist, as well as a dictionary of the data type.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs (hunk 18 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs:21]: Because `PackageSource.ProtocolVersion` is useless, whereas `SourceRepository.GetFeedType` is accurate

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs (hunk 19 above)
Comment (zivkan) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommand/OriginalCaseGlobalPackageFolderTests.cs:26]: In one of the 3 test files, it didn't work for some reason. Since this is defining an alias for a specific type, rather than making a namespace available for use, I thought it isn't so important to keep it with the other usings.

File: src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs (hunk 22 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:718]: When `nuget.config` says that the package source is `.\nupkgs`, this method returns `c:\path\to\repo\.\nupkgs`, and V2 restore logic uses this value, which is then passed to package source telemetry. However, V3 restore logic normalizes the directory name, and creates new `PackageSource` objects, so when V3 sends package source telemetry, the source string does not match.  This changes makes sure it always matches.

File: src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs (hunk 26 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Protocol/RemoteRepositories/HttpFileSystemBasedFindPackageByIdResource.cs:204]: Talked offline. The reason that all the `FindPackageById` methods are instrumented in all implementations is to avoid these types of issues where `GetDependencyInfo` downloads a nupkg in one implementation, but not in another. Only the sum of all resource event times will be comparable, even if `GetDependencyInfo` events are not.
